### PR TITLE
GUARD-3057 Temporarily-Implement-Netco-DoInBatchAsync-ProcessInBatchAsync-in-3dCartAccess

### PR DESCRIPTION
**Pull request by maxim-skuvault**
[GUARD-3057]
Summary: Temporarily implement the Netco DoInBatchAsync and ProcessInBatchAsync in 3dCartAccess

#### Background
Per discussion with Core, we have to stop using Netco. [Chris said it's Ok](https://skuvault.slack.com/archives/CHUENCZ3L/p1692129287319279) to move these methods to Integrations.Core ([TD-270]) and then reference them from this repo ([TD-271]).

#### About these changes
Since we don't want to pause GUARD-3057 until the TD- tickets above, for now we did:
- Implement the Netco DoInBatchAsync, ProcessInBatchAsync in 3dCartAccess
  - Create BatchDetails class to validate method params since our current pattern is constructor validation
- Create unit tests for parameter validation and calls to these new methods

### PR Readiness Checklist

- [ ] I have updated all relevant configuration
- [x] Followed [development conventions][1] to the best of my ability
- [x] Code is documented, particularly public interfaces and hard-to-understand areas
- [x] Tests are updated / added and all pass
- [x] Performed a self-review of my own code
- [x] Acceptance criterias are confirmed by testing changes in UI (or another appropriate way)
- [ ] Confluence documentation is updated, if needed
- [x] New code does not generate new warnings

[1]: https://agileharbor.atlassian.net/wiki/spaces/DEV/pages/1114130/Conventions

[GUARD-3057]: https://agileharbor.atlassian.net/browse/GUARD-3057?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[TD-270]: https://agileharbor.atlassian.net/browse/TD-270?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[TD-271]: https://agileharbor.atlassian.net/browse/TD-271?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/ThreeDCartAccess/Extensions/ParallelProcessExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+				await Task.WhenAll( processingTasks ).ConfigureAwait( false );
+				processingTasks.Clear();
+			}
+		}
+
+		await Task.WhenAll( processingTasks ).ConfigureAwait( false );
+	}
+
+	/// <summary>
+	/// Processes elements asynchronously the in batch of the specified size.
```
Comment (Victoria-SV) [inline src/ThreeDCartAccess/Extensions/ParallelProcessExtensions.cs:49]: small typo. Did you mean "in the batch"?

File: src/ThreeDCartAccess/Extensions/ParallelProcessExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+				AddResultToList( await Task.WhenAll( processingTasks ).ConfigureAwait( false ), result, ignoreNull );
+				processingTasks.Clear();
+			}
+		}
+
+		AddResultToList( await Task.WhenAll( processingTasks ).ConfigureAwait( false ), result, ignoreNull );
+		return result;
+	}
+
+	private static void AddResultToList< TResult >( IEnumerable< TResult > intermidiateResult, List< TResult > endResult, bool ignoreNull )
```
Comment (Victoria-SV) [inline src/ThreeDCartAccess/Extensions/ParallelProcessExtensions.cs:88]: Did you mean intermediateResult? :)

File: src/ThreeDCartAccess/RestApi/ThreeDCartOrdersService.cs [hunk 3]
```diff
@@ -137,7 +137,7 @@ public void GetOrdersByNumber( List< string > invoiceNumbers, DateTime startDate
 		public async Task GetOrdersByNumberAsync( List< string > invoiceNumbers, DateTime startDateUtc, DateTime endDateUtc, Action< ThreeDCartOrder > processAction )
 		{
 			var marker = this.GetMarker();
-			await invoiceNumbers.DoInBatchAsync( 10, async invoiceNumber =>
+			await invoiceNumbers.DoInBatchesAsync( 10, async invoiceNumber =>
```
Comment (Victoria-SV) [inline src/ThreeDCartAccess/RestApi/ThreeDCartOrdersService.cs:140]: Why 10? Looks like a magic number. Could we make it a const?

File: src/ThreeDCartAccess/SoapApi/ThreeDCartOrdersService.cs [hunk 4]
```diff
@@ -118,7 +118,7 @@ public bool IsGetNewOrders( DateTime? startDateUtc = null, DateTime? endDateUtc
 
 		public async Task< List< ThreeDCartOrder > > GetOrdersByNumberAsync( List< string > invoiceNumbers, DateTime startDateUtc, DateTime endDateUtc )
 		{
-			var orders = await invoiceNumbers.ProcessInBatchAsync( 50, async invoiceNumber =>
+			var orders = await invoiceNumbers.DoInBatchesAsync( 50, async invoiceNumber =>
```
Comment (Victoria-SV) [inline src/ThreeDCartAccess/SoapApi/ThreeDCartOrdersService.cs:121]: why 50? Could we make it a const?

File: src/ThreeDCartAccess/Extensions/ParallelProcessExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+	/// <typeparam name="TInput">The type of the input.</typeparam>
+	/// <param name="inputEnumerable">The input enumerable.</param>
+	/// <param name="batchSize">Size of the batch.</param>
+	/// <param name="processor">The processor.</param>
+	/// <returns>Task indicating when all action have been performed.</returns>
+	public static async Task DoInBatchesAsync< TInput >( this IEnumerable< TInput > inputEnumerable, int batchSize, Func< TInput, Task > processor )
+	{
+		var batchDetails = new BatchDetails< TInput >( batchSize, processor );
+
+		await DoInBatchesAsync( inputEnumerable, batchDetails ).ConfigureAwait( false );
```
Comment (Victoria-SV) [inline src/ThreeDCartAccess/Extensions/ParallelProcessExtensions.cs:24]: Do we need ConfigureAwait( false )? I remember we only use it in V1 because V1 it much older.

File: src/ThreeDCartAccessTests/Extensions/BatchDetailsTInputTResultTests.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+	public class BatchDetailsTInputTResultTests
+	{
+		private static readonly Randomizer _randomizer = new Randomizer();
+		private readonly int validBatchSize = ( int )_randomizer.NextUInt( 1, int.MaxValue );
+		private Func< int, Task< int > > validProcessor => i => Task.FromResult( 0 );
+
+		[ TestCase( -1 ) ]
+		[ TestCase( 0 ) ]
+		[ TestCase( null ) ]
+		public void Constructor_ShouldThrow_WhenBatchSizeIsBelowOne( int batchSize )
```
Comment (Victoria-SV) [inline src/ThreeDCartAccessTests/Extensions/BatchDetailsTInputTResultTests.cs:18]: WhenBatchSizeIsBelowOneOrNull would be more accurate

File: src/ThreeDCartAccessTests/Extensions/BatchDetailsTInputTResultTests.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+		}
+
+		[ Test ]
+		public void Constructor_ShouldNotThrow_WhenAllParamsAreValid()
+		{
+			Assert.DoesNotThrow( () => new BatchDetails< int, int >( validBatchSize, validProcessor ) );
+		}
+
+		[ Test ]
+		public void Constructor_ShouldNotThrow_WhenProcessorIsValid_andBatchSizeIsMinimumAllowed()
```
Comment (Victoria-SV) [inline src/ThreeDCartAccessTests/Extensions/BatchDetailsTInputTResultTests.cs:36]: Do we have a maximum allowed? What if a batch size is (int.MaxValue + 1)?

File: src/ThreeDCartAccessTests/Extensions/BatchDetailsTInputTResultTests.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+{
+	public class BatchDetailsTInputTResultTests
+	{
+		private static readonly Randomizer _randomizer = new Randomizer();
+		private readonly int validBatchSize = ( int )_randomizer.NextUInt( 1, int.MaxValue );
+		private Func< int, Task< int > > validProcessor => i => Task.FromResult( 0 );
+
+		[ TestCase( -1 ) ]
+		[ TestCase( 0 ) ]
+		[ TestCase( null ) ]
```
Comment (christian-bejarano-skuvault) [inline src/ThreeDCartAccessTests/Extensions/BatchDetailsTInputTResultTests.cs:17]: @maxim-skuvault Correct me if I'm wrong but it seems we don't need this test case since `batchSize` is an integer and can't be null.

File: src/ThreeDCartAccessTests/Extensions/BatchDetailsTInputTResultTests.cs (hunk 7 above)
Comment (maxim-skuvault) [reply src/ThreeDCartAccessTests/Extensions/BatchDetailsTInputTResultTests.cs:36]: @Victoria-SV There isn't a maximum currently. I copied the code from Netco. Could limit it to batch size of 1000. I think we've run into trouble in production with even size of 10 for popular integrations.

File: src/ThreeDCartAccess/Extensions/ParallelProcessExtensions.cs (hunk 5 above)
Comment (maxim-skuvault) [reply src/ThreeDCartAccess/Extensions/ParallelProcessExtensions.cs:24]: We do use it in internal calls inside *access repos, even in newer ones like amazon-access-v2. So I think it's still a good idea, for contexts to work properly. Won't miss it when these boilerplate calls are no longer needed!

File: src/ThreeDCartAccess/Extensions/ParallelProcessExtensions.cs (hunk 1 above)
Comment (maxim-skuvault) [reply src/ThreeDCartAccess/Extensions/ParallelProcessExtensions.cs]: @Victoria-SV Yes, this is both grammatically incorrect and is somewhat misleading. I clarified it.

File: src/ThreeDCartAccess/Extensions/ParallelProcessExtensions.cs (hunk 2 above)
Comment (maxim-skuvault) [reply src/ThreeDCartAccess/Extensions/ParallelProcessExtensions.cs]: Yes. I copied and couldn't figure why the spell checker was complaining. Corrected

File: src/ThreeDCartAccessTests/Extensions/BatchDetailsTInputTResultTests.cs (hunk 6 above)
Comment (maxim-skuvault) [reply src/ThreeDCartAccessTests/Extensions/BatchDetailsTInputTResultTests.cs:17]: I removed the null case since, [as Christian pointed out](https://github.com/skuvault-integrations/3dCartAccess/pull/15#discussion_r1296300205), it's not possible for an `int`.

File: src/ThreeDCartAccessTests/Extensions/BatchDetailsTInputTResultTests.cs (hunk 8 above)
Comment (maxim-skuvault) [reply src/ThreeDCartAccessTests/Extensions/BatchDetailsTInputTResultTests.cs]: @christian-bejarano-skuvault You're correct. It's not technically possible since `int` isn't nullable. I copied these cases from elsewhere. Turns out for `TestCase( null )` NUnit will just set the batchSize to the default `int` value, which is 0. We already have that case. Removed `[ TestCase( null ) ]` here and in the other test with `batchSize` cases

File: src/ThreeDCartAccess/RestApi/ThreeDCartOrdersService.cs (hunk 3 above)
Comment (maxim-skuvault) [reply src/ThreeDCartAccess/RestApi/ThreeDCartOrdersService.cs:140]: A good idea. I'll create a TODO in the main GUARD-3057 PR https://github.com/skuvault-integrations/3dCartAccess/pull/14 to extract this constant. In this PR just wanted to implement these methods in ThreeDCartAccess.

File: src/ThreeDCartAccess/SoapApi/ThreeDCartOrdersService.cs (hunk 4 above)
Comment (maxim-skuvault) [reply src/ThreeDCartAccess/SoapApi/ThreeDCartOrdersService.cs:121]: @Victoria-SV Same response as above.

**Review by christian-bejarano-skuvault — APPROVED** [review]

**Review by Victoria-SV — APPROVED** [review]
